### ISSUE TITLE
Be able to deploy custom module extension subsystems in swarm

### ISSUE THREAD

**Issue by jmazzitelli** [wildfly-swarm/wildfly-swarm#41]
I have a subsystem that I deploy in WildFly's /modules and add it to standalone.xml via &lt;extension> and &lt;subsystem> and I was wondering how to take that and deploy it in WF-Swarm? There isn't a specific example in WildFly Swarm github that does this (the datasource-subsystem looks close, but I don't think that's it exactly.)

Also, how does one configure their custom subsystem (in WildFly app server installs, you configure it in subsystem's entry inside standalone.xml, but WF-Swarm doesn't have standalone.xml).

But I build using WF-Swarm mvn plugin, my subsystem module jar goes in the swarm /app location and all its dependencies go in m2repo - would this be the  proper setup for module extension subsystems in WF-Swarm or would the mvn plugin need to be enhanced for this type of thing?

In short, I need documentation to show how this can be done (if it can't be done, WF-Swarm should be enhanced to allow it).

**Comment by kenfinnigan** [comment]
It was certainly possible a month or so ago, as I did it for LiveOak, but I haven't tried recently, though there shouldn't be anything preventing it.

For any given WildFly subsystem, you need to create an API and Runtime, as we have for the pieces packaged within WildFly Swarm (such as JPA, Weld, etc). This utilizes the Fraction plugin to build the necessary pieces that can then be consumed by your application.

Not sure if that provides enough detail, but it's definitely possible and something we need to document and provide examples for

**Comment by jmazzitelli** [comment]
ken - you have link to your github repo for the LiveOak work? At minimum, I could use it to see what your mvn layout looked like (what pom dependencies you needed, mvn swarm plugin setup, etc.).

**Comment by kenfinnigan** [comment]
Unfortunately it's all based on Alpha1, which was when a Fraction was a single module, there was no API/Runtime split, and wouldn't be of much help. I've also not pushed it out anywhere yet. I actually need to go back and update it with the new structure.

The main piece is the plugin definition:

        <plugin>
          <groupId>org.wildfly.swarm</groupId>
          <artifactId>wildfly-swarm-fraction-plugin</artifactId>
          <executions>
            <execution>
              <phase>prepare-package</phase>
              <goals>
                <goal>generate</goal>
              </goals>
            </execution>
          </executions>
        </plugin>

In terms of the code, the existing api/runtime modules would be the best example right now.

If you need help with it, I will do my best to help

**Comment by jmazzitelli** [comment]
I'm going to duplicate what is done today in wildfly-swarm/api/datasources and wildfly-swarm/runtime/datasources (by that I mean I'm going to copy what they do in their poms and source structure and java code.) hopefully that gets me close.

**Comment by kenfinnigan** [comment]
it should do, let me know if you hit problems

**Comment by jmazzitelli** [comment]
Ken - I think I have all the code I need, except I dont know if I actually need a Main class. I figure someone has to configure the Fraction, but I can't figure out what I need. I did just build my swarm jar and I got this error:

Caused by: java.util.ServiceConfigurationError: org.wildfly.swarm.runtime.container.ServerConfiguration: Provider org.wildfly.swarm.runtime.connector.ConnectorConfiguration not a subtype

So I have a feeling I am missing some dependencies in my poms.

Note - the code is here (there are three mvn modules - api, runtime, and dist - the "dist" one is where I run the swarm plugin to build the fatjar): https://github.com/hawkular/hawkular-agent/tree/mazz/wf9-swarm/hawkular-agent-swarm

P.S. Is this the right forum to discuss this? (github issues) or is there a swarm mailing list to use?

**Comment by bobmcwhirter** [comment]
There is a series of service locators to find and auto configure your fraction if no main is provided. And to find the runtime module. It's tangly but works. 

--

**Comment by jmazzitelli** [comment]
Ok, I see a call to RuntimeServer.applyDefaults ... so I'll assume I'll get something to start my service.

But I'm getting this and not sure why:

Caused by: java.util.ServiceConfigurationError: org.wildfly.swarm.runtime.container.ServerConfiguration: Provider org.wildfly.swarm.runtime.connector.ConnectorConfiguration not a subtype
	at java.util.ServiceLoader.fail(ServiceLoader.java:239)
	at java.util.ServiceLoader.access$300(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:376)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at org.wildfly.swarm.runtime.container.RuntimeServer.applyFractionDefaults(RuntimeServer.java:106)
	at org.wildfly.swarm.runtime.container.RuntimeServer.applyDefaults(RuntimeServer.java:67)
	at org.wildfly.swarm.runtime.container.RuntimeServer.start(RuntimeServer.java:51)
	at org.wildfly.swarm.container.Container.start(Container.java:112)
	at org.wildfly.swarm.Swarm.main(Swarm.java:12)

**Comment by bobmcwhirter** [comment]
I think you might’ve copied this file verbatim:

	https://github.com/wildfly-swarm/wildfly-swarm/blob/master/runtime/connector/src/main/resources/META-INF/services/org.wildfly.swarm.runtime.container.ServerConfiguration <https://github.com/wildfly-swarm/wildfly-swarm/blob/master/runtime/connector/src/main/resources/META-INF/services/org.wildfly.swarm.runtime.container.ServerConfiguration>

And instead it should list your WhateverConfiguration class in your part of the tree.

Or your copy of this file:  

	https://github.com/wildfly-swarm/wildfly-swarm/blob/master/api/connector/src/main/resources/META-INF/services/org.wildfly.swarm.container.RuntimeModuleProvider <https://github.com/wildfly-swarm/wildfly-swarm/blob/master/api/connector/src/main/resources/META-INF/services/org.wildfly.swarm.container.RuntimeModuleProvider>

isn’t pointing to your runtime module correctly.

Container loads the API, then service-locators up the RuntimeModuleProvider from the API half of the works.  It then queries it for the name of the jboss-module for the runtime half of the works, loads it, and then asks that module for a service-locator for the appropriate ServerConfiguration implementations.

So, there’s two META-INF/services/ you have to adjust when you copy/paste.  One in the API half, and one in the runtime half.  Either one could eventually result in what you’re seeing.

Maybe.

	-Bob

**Comment by kenfinnigan** [comment]
@jmazzitelli I'm currently working on your Fraction. Will send a PR when I know it works

**Comment by jmazzitelli** [comment]
Those look OK in my project - no copy-paste error. These point to my classes:

https://github.com/hawkular/hawkular-agent/blob/mazz/wf9-swarm/hawkular-agent-swarm/hawkular-agent-swarm-api/src/main/resources/META-INF/services/org.wildfly.swarm.container.RuntimeModuleProvider

https://github.com/hawkular/hawkular-agent/blob/mazz/wf9-swarm/hawkular-agent-swarm/hawkular-agent-swarm-runtime/src/main/resources/META-INF/services/org.wildfly.swarm.runtime.container.ServerConfiguration

**Comment by kenfinnigan** [comment]
@jmazzitelli I've fixed the problems with the Hawkular Fraction here: https://github.com/hawkular/hawkular-agent/pull/5

However, the Hawkular Subsystem is coded against WildFly 8.2.0.Final, but WildFly Swarm relies on a minimum of 9.0.0.CR1. Until the Hawkular Subsystem is updated to work with WildFly 9.0.0.CR1, WildFly Swarm won't work with it.
